src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/PolymorphicTypeResolver.cs

```diff
@@ -0,0 +1,252 @@
+
+        public bool TryGetDerivedJsonTypeInfo(Type runtimeType, [NotNullWhen(true)] out JsonTypeInfo? jsonTypeInfo, out string? typeDiscriminatorId)
+        {
+            Debug.Assert(BaseType.IsAssignableFrom(runtimeType));
+
+            if (!_typeToDiscriminatorId.TryGetValue(runtimeType, out DerivedJsonTypeInfo? result))
+            {
+                switch (UnknownDerivedTypeHandling)
+                {
+                    case JsonUnknownDerivedTypeHandling.FallbackToNearestAncestor:
+                        // Calculate (and cache the result) of the nearest ancestor for given runtime type.
+                        // A `null` result denotes no matching ancestor type, we also cache that.
+                        result = CalculateNearestAncestor(runtimeType);
+                        _typeToDiscriminatorId[runtimeType] = result;
+                        break;
+                    case JsonUnknownDerivedTypeHandling.FallbackToBaseType:
+                        // Recover the polymorphic contract (i.e. any type discriminators) for the base type, if it exists.
+                        _typeToDiscriminatorId.TryGetValue(BaseType, out result);
+                        _typeToDiscriminatorId[runtimeType] = result;
+                        break;
+
+                    case JsonUnknownDerivedTypeHandling.FailSerialization:
+                    default:
+                        if (runtimeType != BaseType)
+                        {
+                            ThrowHelper.ThrowNotSupportedException_RuntimeTypeNotSupported(BaseType, runtimeType);
+                        }
+                        break;
+                }
+            }
+
+            if (result is null)
+            {
+                jsonTypeInfo = null;
+                typeDiscriminatorId = null;
+                return false;
+            }
+            else
+            {
+                jsonTypeInfo = result.GetJsonTypeInfo(_options);
+                typeDiscriminatorId = result.TypeDiscriminatorId;
+                return true;
+            }
+        }
+
+        public bool TryGetDerivedJsonTypeInfo(string typeDiscriminatorId, [NotNullWhen(true)] out JsonTypeInfo? jsonTypeInfo)
+        {
+            Debug.Assert(UsesTypeDiscriminators);
+            Debug.Assert(_discriminatorIdtoType != null);
+
+            if (_discriminatorIdtoType.TryGetValue(typeDiscriminatorId, out DerivedJsonTypeInfo? result))
+            {
+                Debug.Assert(result.TypeDiscriminatorId == typeDiscriminatorId);
+                jsonTypeInfo = result.GetJsonTypeInfo(_options);
+                return true;
+            }
+
+            if (!IgnoreUnrecognizedTypeDiscriminators)
+            {
+                ThrowHelper.ThrowJsonException_UnrecognizedTypeDiscriminator(typeDiscriminatorId);
+            }
+
+            jsonTypeInfo = null;
+            return false;
+        }
+
+        public static bool IsSupportedPolymorphicBaseType(Type? type) =>
```

Comment:
nit: this method is currently not called with an arg typed as nullable.

```suggestion
        public static bool IsSupportedPolymorphicBaseType(Type type) =>
```

Reply:
Added this intentionally to emphasize the scope of validation. In any case it's an internal helper method so it shouldn't matter very much.